ethereum/executionlayer/src/main/java/tech/pegasys/teku/ethereum/executionlayer/ExecutionLayerManagerImpl.java

```diff
@@ -247,23 +281,80 @@ public SafeFuture<ExecutionPayloadHeader> builderGetHeader(
+            builderBidV1SignedMessage -> getHeaderFromBuilderBid(builderBidV1SignedMessage, slot))
         .thenPeek(
-            executionPayloadHeader ->
-                LOG.trace(
-                    "builderGetHeader(slot={}, pubKey={}, parentHash={}) -> {}",
-                    slot,
-                    Bytes48.ZERO,
-                    executionPayloadContext.getParentHash(),
-                    executionPayloadHeader));
+            executionPayloadHeader -> {
+              // store that we haven't fallen back for this slot
+              slotToLocalElFallbackData.put(slot, Optional.empty());
+              LOG.trace(
+                  "builderGetHeader(slot={}, pubKey={}, parentHash={}) -> {}",
+                  slot,
+                  Bytes48.ZERO,
+                  executionPayloadContext.getParentHash(),
+                  executionPayloadHeader);
+            })
+        .exceptionallyCompose(
+            error -> {
+              LOG.error(
+                  "builderGetHeader returned an error. Falling back to local execution engine",
+                  error);
+              return doFallbackToLocal(localExecutionPayload, slot);
+            });
+  }
+
+  @Override
+  public SafeFuture<ExecutionPayload> builderGetPayload(
+      final SignedBeaconBlock signedBlindedBeaconBlock) {
+
+    checkArgument(
+        signedBlindedBeaconBlock.getMessage().getBody().isBlinded(),
+        "SignedBeaconBlock must be blind");
+
+    final UInt64 slot = signedBlindedBeaconBlock.getSlot();
+
+    final Optional<Optional<SafeFuture<ExecutionPayload>>> maybeFallbackPayload =
+        Optional.ofNullable(slotToLocalElFallbackData.get(slot));
+
+    if (maybeFallbackPayload.isEmpty()) {
+      LOG.warn(
+          "Blinded block seems not been built via either builder or local engine. Trying to unblind it via builder endpoint anyway.");
+      return getPayloadFromBuilder(signedBlindedBeaconBlock);
+    }
+
+    if (maybeFallbackPayload.get().isEmpty()) {
+      return getPayloadFromBuilder(signedBlindedBeaconBlock);
+    }
+
+    slotToLocalElFallbackData.remove(slot);
+
+    // fallback to local execution engine
+    return maybeFallbackPayload.get().get();
+  }
+
+  private SafeFuture<ExecutionPayloadHeader> doFallbackToLocal(
+      final SafeFuture<ExecutionPayload> localExecutionPayload, final UInt64 slot) {
+
+    return localExecutionPayload
+        .thenApply(
+            executionPayload ->
+                spec.atSlot(slot)
+                    .getSchemaDefinitions()
+                    .toVersionBellatrix()
+                    .orElseThrow()
+                    .getExecutionPayloadHeaderSchema()
+                    .createFromExecutionPayload(executionPayload))
+        .thenPeek(
+            __ -> {
+              // store the fallback payload for this slot,
+              slotToLocalElFallbackData.put(slot, Optional.of(localExecutionPayload));
```

Comment:
It feels a little weird that we're storing a `SafeFuture` here but only after it is completed. I'm not sure if that's opening a potential race condition and we should be storing the SafeFuture before it completes or if we could remove ambiguity by storing the executionPayload rather than `SafeFuture<ExecutionPayload>` - even though we'll wind up wrapping it in `completedFuture` when we wind up getting it out of the map again it makes the logic clearer that if it's there it's complete.

Reply:
oops, right. It was my last change to `put` in the `peek` and I haven't realize it.